tracer/src/Datadog.Trace/AppSec/Waf/Waf.cs

```diff
@@ -66,53 +108,64 @@ public void Dispose(bool disposing)
 
             disposed = true;
 
-            rule?.Dispose();
+            wafHandle?.Dispose();
         }
 
-        private static Obj RuleSetFromManifest(List<Obj> argCache)
+        private static Obj CreatObjFromRulesStream(List<Obj> argCache, Stream stream)
         {
-            var assembly = typeof(Waf).Assembly;
-            var resource = assembly.GetManifestResourceStream("Datadog.Trace.AppSec.Waf.rule-set.json");
-            using var reader = new JsonTextReader(new StreamReader(resource));
-            var root = JToken.ReadFrom(reader);
+            using var reader = new StreamReader(stream);
+            using var jsonReader = new JsonTextReader(reader);
+            var root = JToken.ReadFrom(jsonReader);
+
+            LogRuleDetailsIfDebugEnabled(root);
 
             return Encoder.Encode(root, argCache);
         }
 
-        private static WafHandle NewRule()
+        private static Stream GetRulesManifestStream()
         {
-            try
-            {
-                DdwafConfigStruct args = default;
-
-                var argCache = new List<Obj>();
-                var rules = RuleSetFromManifest(argCache);
+            var assembly = typeof(Waf).Assembly;
+            return assembly.GetManifestResourceStream("Datadog.Trace.AppSec.Waf.rule-set.json");
+        }
 
-                var ruleHandle = WafNative.Init(rules.RawPtr, ref args);
+        private static Stream GetRulesFileStream(string rulesFile)
+        {
+            if (!File.Exists(rulesFile))
+            {
+                Log.Error("AppSec could not find the rules file in path \"{RulesFile}\". AppSec will not run any protections in this application.", rulesFile);
+                return null;
+            }
 
-                rules.Dispose();
-                foreach (var arg in argCache)
-                {
-                    arg.Dispose();
-                }
+            return File.OpenRead(rulesFile);
+        }
 
-                if (ruleHandle == IntPtr.Zero)
+        private static void LogRuleDetailsIfDebugEnabled(JToken root)
+        {
+            if (Log.IsEnabled(LogEventLevel.Debug))
+            {
+                try
                 {
-                    Log.Error("Failed to create rules.");
-                    return null;
+                    var eventsProp = root.Value<JArray>("events");
+                    foreach (var ev in eventsProp)
+                    {
+                        var idProp = ev.Value<JValue>("id");
+                        var nameProp = ev.Value<JValue>("name");
+                        var addresses = ev.Value<JArray>("conditions").SelectMany(x => x.Value<JObject>("parameters").Value<JArray>("inputs"));
+                        Log.Debug($"Loaded rule: {idProp.Value} - {nameProp.Value} on addresses: {string.Join(", ", addresses)}");
```

Comment:
String allocation is not an issue here because we check if debug is enabled, but we still want the structured logging:
```suggestion
                        Log.Debug("Loaded rule: {id} - {name} on addresses: {addresses}", idProp.Value, nameProp.Value, string.Join(", ", addresses));
```

Reply:
okay, but why?

Reply:
2 reasons come to mind:
* Serilog (which we're using under the covers) uses the message template to generate an EventId. I can't remember if it uses that for caching, but the assumption is that the message template is constant
* If we decide to output the logs to another sink (which we could feasibly do in the future, and which some customers have requested) then we should output using structured logging where available, to ensure we have a constant message template hash, and to enable easier searching/grouping by message properties